autoprecompiles/src/constraint_optimizer.rs

```diff
@@ -88,6 +93,97 @@ fn solver_based_optimization<T: FieldElement, V: Clone + Ord + Hash + Display>(
     Ok(constraint_system)
 }
 
+/// Removes free variables from the constraint system, under some conditions.
+///
+/// Motivation: Suppose there is a constraint `2 * foo = bar` and `foo` only appears in this constraint.
+/// Then, if we assume that all constraints are satisfiable, the prover would be able to satisfy it for
+/// any value of `bar` by solving for `foo`. Therefore, the constraint can be removed.
+/// The same would be true for a *stateless* bus interaction, e.g. `[foo * bar] in [BYTES]`.
+///
+/// This function *some* constraints like this (see TODOs below).
+fn remove_free_variables<T: FieldElement, V: Clone + Ord + Eq + Hash + Display>(
+    mut constraint_system: JournalingConstraintSystem<T, V>,
+    solver: &mut impl Solver<T, V>,
+    bus_interaction_handler: impl IsBusStateful<T> + Clone,
+) -> JournalingConstraintSystem<T, V> {
+    let all_variables = constraint_system
+        .system()
+        .expressions()
+        .flat_map(|expr| expr.referenced_variables())
+        .cloned()
+        .collect::<HashSet<_>>();
+
+    let variables_to_delete = all_variables
+        .iter()
+        // Find variables that are referenced in exactly one constraint
+        .filter_map(|variable| {
+            constraint_system
+                .indexed_system()
+                .constraints_referencing_variables(once(variable.clone()))
+                .exactly_one()
+                .ok()
+                .map(|constraint| (variable.clone(), constraint))
+        })
+        .filter(|(variable, constraint)| match constraint {
+            // TODO: These constraints could be removed also if they are linear in the free variable.
+            // The problem with this currently is that this removes constraints like
+            // `writes_aux__prev_data__3_0 - BusInteractionField(15, 7)` (`writes_aux__prev_data__3_0` is a free variable)
+            // which causes `remove_bus_interaction_variables` to fail, because it doesn't know the definition of the
```

Comment:
I hope I can ditch this in one of the next PRs. Can you turn this into an issue?

Reply:
#3113